### ISSUE TITLE
RouteBase.EnsureLoggers race condition

### ISSUE THREAD

**Issue by brettsam** [aspnet/AspNetCore#11183]
### Describe the bug
I'm on the Azure Functions team and we see a very rare NullRef with this stack:
```
System.ArgumentNullException : Value cannot be null.
Parameter name: logger
   at Microsoft.AspNetCore.Routing.RouteConstraintMatcher.Match(IDictionary`2 constraints,RouteValueDictionary routeValues,HttpContext httpContext,IRouter route,RouteDirection routeDirection,ILogger logger)
   at Microsoft.AspNetCore.Routing.RouteBase.RouteAsync(RouteContext context)
```

I've looked into the issue (I thought it was a race on our end) but found that `_constraintLogger` (in `RouteBase`) ends up being null, even with the call to `EnsureLoggers()` that should be making sure it's initialized. 

```
private void EnsureLoggers(HttpContext context)
{
    if (_logger == null)
    {
        var factory = context.RequestServices.GetRequiredService<ILoggerFactory>();
        _logger = factory.CreateLogger(typeof(RouteBase).FullName);
        _constraintLogger = factory.CreateLogger(typeof(RouteConstraintMatcher).FullName);
    }
}
```

Imagine if the second `CreateLogger()` call here runs slowly (for any reason: Functions runs on a single core, file locks, etc -- it’d be rare, but possible). Now you’ve got `_logger != null`, but `_constraintLogger == null`. Any other threads calling this code will skip over this `if` statement and go on with their execution -- with a null `_constraintLogger`. If those two `CreateLogger()` lines were switched, it should fix it.

### To Reproduce
See my test -- required some sempahore coordination, but proved that it's possible, but rare: https://gist.github.com/brettsam/6f60bfccba80d1515be9cfb13d0d985b

**Comment by Tratcher** [comment]
@JamesNK @rynowak

**Comment by rynowak** [comment]
@brettsam - thanks for the thorough bug report. This is indeed a bug. 

@mkArtakMSFT - we should consider a backport of this, it's a low risk fix for a functional bug.

**Comment by mkArtakMSFT** [comment]
@rynowak is this fixed in 3.0 already? Or are you referring to whatever that fix would be we should backport it?

**Comment by rynowak** [comment]
Not fix yet, but it should be very easy.

**Comment by javiercn** [comment]
I do have a PR for this out already.

**Comment by anurse** [comment]
> we should consider a backport of this, it's a low risk fix for a functional bug.

FYI, after conversations with the Functions team, this is a **very rare** occurrence and they do not feel they need a backport. If it's something we still want to backport, that's fine.

**Comment by javiercn** [comment]
Fixed in 3.0
